Add authorization code flow extAuth tutorial (#2387)
kyma-project/api-gateway: docs/user/tutorials/01-54-secure-workload-extauth-authorization-code-flow.md

## Procedure

### Create and Configure an OpenID Connect Application

In SAP Cloud Identity Services, create an OpenID Connect application and configure it for the Authorization Code flow.

1. Sign in to the administration console for SAP Cloud Identity Services. See [Access Admin Console](https://help.sap.com/docs/cloud-identity-services/cloud-identity-services/accessing-administration-console?locale=en-US&version=Cloud).

2. Create an OpenID Connect Application.

   1. Go to **Application Resources** > **Application**.
   2. Choose **Create**, provide the application name, and select the OpenID Connect radio button. 
      For more configuration options, see [Create OpenID Connect Application](https://help.sap.com/docs/cloud-identity-services/cloud-identity-services/create-openid-connect-application-1a87534329494d48ae5f246c4842e11a?locale=en-US&version=Cloud).
   3. Choose **+Create**.

3. Configure OpenID Connect Application for the Authorization Code flow.
   
   1. In the **Trust > Single Sign-On** section of your created application, choose **OpenID Connect Configuration**.
   2. Provide the name.

>    3. Add the `https://oauth2-proxy.{YOUR_DOMAIN}/oauth2/callback` redirect URI. 

Reviewer: ```suggestion
   3. In the redirect URIs section add the `https://oauth2-proxy.{YOUR_DOMAIN}/oauth2/callback`. 
```
i think it is more readable that way

              include:
                - "authorization"
                - "cookie"
                - "x-forwarded-for"
              add:
                # Sets uri to redirect user after login
                x-forwarded-uri: "https://%REQ(:authority)%%REQ(x-envoy-original-path?:path)%"
            toUpstream:
              onAllow:
              - "authorization"
              - "cookie"
              - "path"
              - "x-forwarded-id-token"
              - "x-forwarded-access-token"
              - "x-forwarded-refresh-token"
              - "x-forwarded-for"
              - "x-forwarded-user"
              - "x-forwarded-groups"
              - "x-forwarded-email"
              - "x-forwarded-preferred-username"
            toDownstream:
              onDeny:
              - "content-type"
              - "set-cookie"
              onAllow:
              - "x-forwarded-id-token"
              - "x-forwarded-access-token"
              - "x-forwarded-refresh-token"
              - "x-forwarded-for"
              - "x-forwarded-user"
              - "x-forwarded-groups"
              - "x-forwarded-email"
              - "x-forwarded-preferred-username"
    EOF
    ```

> 

Reviewer: Maybe we can link here istio cr documentation 
sth like:
```suggestion
  For more information about fields set above see  [Istio Custom Resource documentation](https://kyma-project.io/external-content/istio/docs/user/04-00-istio-custom-resource.html).
```

# Expose and Secure a Workload with OAuth2 Proxy External Authorizer (Authorization Code Flow)
Learn how to expose and secure a workload using OAuth2 Proxy external authorizer and the OAuth 2.0 Authorization Code flow. SAP Cloud Identity Services acts as the OAuth2/OIDC Identity Provider (IdP) that authenticates users.

## Prerequisites

- You have an SAP BTP, Kyma runtime instance with the Istio and API Gateway modules added. The Istio and API Gateway modules are added to your Kyma cluster by default.
- You have an SAP Cloud Identity Services tenant. See [Initial Setup](https://help.sap.com/docs/cloud-identity-services/cloud-identity-services/initial-setup?locale=en-US&version=Cloud&q=open+id+connect).
- You have installed [helm](https://helm.sh/docs/intro/install).

## Context

This procedure shows how to implement external authorization for your Kyma workloads using the OAuth 2.0 Authorization Code flow.

> When the user visits an URL of your exposed workload, the following steps take place:

Reviewer 2: ```suggestion
When the user visits a URL of your exposed workload, the following steps take place:
```

...
```
<!-- tabs:end -->

See the following example APIRule with **extAuth** authorizer that exposes the sample HTTPBin Deployment:

1. Create the `httpbin-system` namespace and deploy a sample HTTPBin Deployment.

    ```bash
    kubectl create ns httpbin-system
    kubectl label namespace httpbin-system istio-injection=enabled
    kubectl apply -f \
    https://raw.githubusercontent.com/istio/istio/master/samples/httpbin/httpbin.yaml \
    -n httpbin-system
    ```

2. Expose the workload with an APIRule using the extAuth access strategy. 

    ```bash
    cat <<EOF | kubectl apply -f -
    apiVersion: gateway.kyma-project.io/v2
    kind: APIRule
    metadata:
      name: httpbin
      namespace: httpbin-system
    spec:
      gateway: ${GATEWAY}
      hosts:
        - httpbin.${EXPOSE_DOMAIN}
      service:
        name: httpbin
        port: 8000
      rules:
        - path: /*
          methods: ["GET"]
          extAuth:
            authorizers:
              - oauth2-proxy
    EOF
    ```
    
    Check if the APIRule's status is ready:
    
    ```bash
    kubectl --namespace=oauth2-proxy get apirules httpbin -n httpbin-system
    ```

#### Results

> To access your workload copy the host linked defined in your APIRule and open it in a browser. You're redirected to Cloud Identity Services first where you need to log in. If the login is successful, you can access your application. In the sample scenario, when you open `httpbin.${EXPOSE_DOMAIN}`, the page displays HTTPBin endpoints.

Reviewer 2: ```suggestion
To access your workload, copy the host linked defined in your APIRule and open it in a browser. You're redirected to Cloud Identity Services first, where you need to log in. If the login is successful, you can access your application. In the sample scenario, when you open `httpbin.${EXPOSE_DOMAIN}`, the page displays HTTPBin endpoints.
```

Author: the host link defined, thank you
